src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey.c

```diff
@@ -89,7 +89,220 @@ int32_t CryptoNative_UpRefEvpPkey(EVP_PKEY* pkey)
+    if (BN_is_zero(n))
+    {
+        ERR_put_error(ERR_LIB_EVP, 0, EVP_R_DECODE_ERROR, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // OpenSSL has kept this value at 16,384 for all versions.
+    if (BN_num_bits(n) > OPENSSL_RSA_MAX_MODULUS_BITS)
+    {
+        ERR_PUT_error(ERR_LIB_RSA, 0, RSA_R_MODULUS_TOO_LARGE, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // Exponent cannot be 1 and must be odd
+    if (BN_is_one(e) || !BN_is_odd(e))
+    {
+        ERR_PUT_error(ERR_LIB_RSA, 0, RSA_R_BAD_E_VALUE, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // At this point everything that is public has been checked. Mark as successful and clean up.
+    if (isPublic)
+    {
+        ret = 1;
+        goto done;
+    }
+
+    // We do not support validating multi-prime RSA. If there are extra primes (more than two) then treat it as a
+    // decoding failure.
+    if (RSA_get_multi_prime_extra_count(rsa) != 0)
+    {
+        ERR_put_error(ERR_LIB_EVP, 0, EVP_R_DECODE_ERROR, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // Get the private components now that we've moved on to checking the private parameters.
+    RSA_get0_factors(rsa, &p, &q);
+
+    // Need all the private parameters now.
+    if (!d || !p || !q)
+    {
+        ERR_PUT_error(ERR_LIB_RSA, 0, RSA_R_VALUE_MISSING, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // Set up the scratch integers
+    if ((ctx = BN_CTX_new()) == NULL ||
+        (x = BN_new()) == NULL ||
+        (y = BN_new()) == NULL ||
+        (p1 = BN_new()) == NULL ||
+        (q1 = BN_new()) == NULL)
+    {
+        goto done;
+    }
+
+    // multiply p and q and put the result in x.
+    if (!BN_mul(x, p, q, ctx))
+    {
+        goto done;
+    }
+
+    // p * q == n
+    if (BN_cmp(x, n) != 0)
+    {
+        ERR_PUT_error(ERR_LIB_RSA, 0, RSA_R_N_DOES_NOT_EQUAL_P_Q, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // Checking congruence of private parameters.
+    // de = 1 % lambda(n)
+    // lambda(n) = lcm(p-1, q-1)
+    // lambda(n) is known to be lambda(pq) already.
+    // p1 = p-1
+    // q1 = q-1
+    // x = lcm(x, y)
+    if (!BN_sub(p1, p, BN_value_one()) || !BN_sub(q1, q, BN_value_one()) || !Lcm(p1, q1, ctx, x))
+    {
+        goto done;
+    }
+
+    // de % lambda(n)
+    // put the result in y
+    if (!BN_mod_mul(y, d, e, x, ctx))
+    {
+        goto done;
+    }
+
+    // Given de = 1 % lambda(n), de % lambda(n) should be one.
+    if (!BN_is_one(y))
+    {
+        ERR_PUT_error(ERR_LIB_RSA, 0, RSA_R_D_E_NOT_CONGRUENT_TO_1, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // Move on to checking the CRT parameters. In compatibility with what OpenSSL does,
+    // these are optional and only check them if all are present.
+    RSA_get0_crt_params(rsa, &dp, &dq, &inverseQ);
+
+    if (dp && dq && inverseQ)
+    {
+        // Check dp = d % (p-1)
+        // compute d % (p-1) and put in x
+        if (!BN_div(NULL, x, d, p1, ctx))
```

Comment:
I'm _pretty_ sure we have both d-phi and d-lambda keys in our tests, so we probably already have test coverage, but I stopped and thought so I'm sharing my thoughts:

I believe that this code is only checking that the CRT parameters are correct for the specified D; and otherwise that the specified D is _a_ valid D (not "the best" D).  So we're not having an opinion of "you didn't produce the D/DP/DQ/QInv values that I expected".

It... might... be worth a comment to that effect?

Reply:
OpenSSL does not check that you chose the "the best" D, only a "working" D. When OpenSSL is in FIPS mode, it does do that check (SP800-56Br1 6.4.1.2.3). I did not want to make the check tighter than what is more typically used.

But yes, we can add a comment.